tests/performance/density-kwok.go

```diff
@@ -0,0 +1,305 @@
+package performance
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/pkg/libvmi"
+	"kubevirt.io/kubevirt/tests/framework/kubevirt"
+	"kubevirt.io/kubevirt/tests/libvmifact"
+	"kubevirt.io/kubevirt/tests/util"
+)
+
+const (
+	vmBatchStartupLimit = 5 * time.Minute
+	defaultNodeCount    = 100
+	defaultVMCount      = 1000
+)
+
+var (
+	nodeCount = getNodeCount()
+	vmCount   = getVMCount()
+)
+
+var _ = SIGDescribe("[KWOK] Control Plane Performance Density Testing using kwok", func() {
+	var (
+		kubevirtClient kubecli.KubevirtClient
+		k8sClient      *kubernetes.Clientset
+		startTime      time.Time
+	)
+
+	artifactsDir, _ := os.LookupEnv("ARTIFACTS")
+
+	BeforeEach(func() {
+		skipIfNoKWOKPerformanceTests()
+		startTime = time.Now()
+	})
+
+	Describe("kwok density tests", func() {
+		Context(fmt.Sprintf("create a batch of %d fake Nodes", nodeCount), func() {
+			It("should successfully create fake nodes", func() {
+				kubevirtClient = kubevirt.Client()
+
+				config, err := kubecli.GetKubevirtClientConfig()
+				if err != nil {
+					fmt.Fprintf(os.Stderr, "failed to get client config: %v\n", err)
+					return
+				}
+
+				k8sClient, err = kubernetes.NewForConfig(config)
+				if err != nil {
+					fmt.Fprintf(os.Stderr, "failed to create k8s client: %v\n", err)
+					panic(err)
+				}
+
+				By("create fake Nodes")
+				createFakeNodesWithKwok(k8sClient, nodeCount)
+
+				By("Get the list of nodes")
+				_, err = k8sClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+				if err != nil {
+					log.Fatalf("Failed to list nodes: %v", err)
+				}
+			})
+		})
+
+		Context(fmt.Sprintf("create a batch of %d fake VMIs", vmCount), func() {
+			It("should sucessfully create all fake VMIs", func() {
+				By("Creating a batch of fake VMIs")
+				createFakeVMIBatchWithKWOK(kubevirtClient, vmCount)
+
+				By("Waiting for a batch of VMIs")
+				waitRunningVMI(kubevirtClient, vmCount, vmBatchStartupLimit)
+
+				By("Deleting fake VMIs")
+				deleteAndVerifyFakeVMIBatch(kubevirtClient, vmCount, vmBatchStartupLimit)
+
+				By("Collecting metrics")
+				collectMetrics(startTime, filepath.Join(artifactsDir, "VMI-kwok-perf-audit-results.json"))
+
+				//wait for 5 mins to bring the metrics to steady state
+				time.Sleep(5 * time.Minute)
```

Comment:
Isn't it better to wait for the steady state looking the metric itself instead of sleeping?

Reply:
And why in the following test (`should sucessfully create all fake VMs`) this sleep is not present?

Reply:
The idea behind adding the sleep time is that creation and deletion will spike up the resource utilization which will impact the subsequent test. So, waiting for sometime after will bring the metrics to the steady state and metrics collected in each context will be more appropriate.

> And why in the following test (should sucessfully create all fake VMs) this sleep is not present?
We are not collecting the metrics after this context, so sleep is not present.

Reply:
Is there a way to wait for it to be done without waiting a fixed amount of time? E.g. monitoring a condition?

Reply:
One way is to check if CPU usage and memory usage are in the acceptable ranges or not. Can you help me understand what are the acceptable ranges or baseline values.
cc: @alaypatel07

Reply:
Empirically it is observed that creating and deleting 1000s of VMIs will spike up resource utilization.  Since one test is immediately run after another, it would be good to not have the previous test deletion metric impact the next test creation. At the same time it would be really hard to have a more dynamic sanity check as to when the steady state is achieved.

For this reason, can we start with simply sleeping for 1 min i.e. 2 x PrometheusScrapeInterval after each test?

Reply:
As discussed, added primer test and reduced the sleeping time to  2 x PrometheusScrapeInterval in aftereach block